core/pool_vector.h

```diff
@@ -411,8 +411,8 @@ class PoolVector {
 			p_to = size() + p_to;
 		}
 
-		CRASH_BAD_INDEX(p_from, size());
-		CRASH_BAD_INDEX(p_to, size());
+		ERR_FAIL_INDEX_V(p_from, size(), PoolVector<T>());
+		ERR_FAIL_INDEX_V(p_to, size(), PoolVector<T>());
```

Comment:
It seems those were made crashing on purpose by @RandomShaper in 211c4518903d82068c061943064824ac5595fd38. It would be good to know why, and assess the other `CRASH_BAD_INDEX` added back then.

Reply:
Before my commit, the error handling code was this:
```
PoolVector<T> &aux = *((PoolVector<T> *)0); // nullreturn
ERR_FAIL_COND_V(p_from < 0 || p_from >= size(), aux)
```
Seemingly, the intention was to make the engine crash at some point, but that way of doing it invokes undefined behavior. It might very well keep running on a weird state.

I made the crash explicit and happening for sure.

This new change modifies the behavior of the function so that it handles out-of-bounds indices gracefully, giving you back and empty vector instead of crashing. But I'm not sure if that would be losing a chance of catching a logic error early.

Reply:
Thanks for the explanation, I was actually unsure about this change but forgot to check the log and just tagged hpvb since I knew he worked on that time ago.

> But I'm not sure if that would be losing a chance of catching a logic error early.

I think the idea in Godot is to always avoid crashing, and let the user take responsibility in not having errors in the debugger/console. So, in case of a logic error, the user would see the message and realize that something is wrong when calling `resize`. I think the editor also point you to the line in code that called if if you click on it. But maybe I'm missing the point.

Reply:
But maybe we can handle this at a higher level in the exposed functions? Would that be better? I find it strange that passing a negative value to `resize` in gdscript crashes the engine.

Reply:
Yeah I don't think it should crash if it is recoverable, it should just error out. Red errors printed to the stdout are still must fix for both game developers and engine devs, so it's sufficient IMO.

Reply:
I totally agree on your point about `resize`.
Other crashing functions can not be fixed so easily; i.e., those returning references to elements.
I find your idea about handling those situations more gracefully at the binding level interesting.
Some methods 
What if we undertake a full review of all the crashing behaviors for 4.0?
Not sure if a more conservative subset of that would have a point in 3.2 if we did that.

Reply:
> What if we undertake a full review of all the crashing behaviors for 4.0?

Then you might want to look at #24199. FAIL macros don't print the error line in script, only the line in the source, which in many cases isn't really useful. I wonder if it's possible to give macro an access to the script, so it tells you which line in YOUR code fails.

Reply:
Comment from @reduz on IRC:
```
18:53 <reduz> Akien: since this can be used from gdscript, it may probably be better to throw an error and return an empty vector
18:53 <reduz> crashing is too much
18:53 <reduz> from gdscript all functions that can do out of bounds reads are prevalidated
18:53 <reduz> (and writes)
```

Reply:
@KoBeWi don't the editor debugger bring you to the GDScript line that caused that error when you click on it? It should.

Reply:
See the attached this [screencast](https://github.com/godotengine/godot/files/3321911/debugger_to_line2.zip) as an example

Reply:
I remember some errors that you couldn't trace, but maybe they weren't from GDScript, but from scene (in which case providing a failing node path would be helpful).